File: 2015/3/3-2.py
def move(x, y, c):
  if c == '>':
    x += 1
  elif c == '<':
    x -= 1
  elif c == '^':
    y += 1
  elif c == 'v':
    y -= 1
  return (x, y)

with open('input.txt') as f:
  path = f.readline().strip('\n')

(s_x, s_y) = [0, 0]
(r_x, r_y) = [0, 0]
santa = True

visits = {}


for c in path:
  if santa:
    visits[(s_x, s_y)] = True
    (s_x, s_y) = move(s_x, s_y, c)
    visits[(s_x, s_y)] = True
    santa = False
  else:
    visits[(r_x, r_y)] = True
    (r_x, r_y) = move(r_x, r_y, c)
    visits[(r_x, r_y)] = True
    santa = True

print('Houses Visited: %d' % (len(visits.keys())))

File: 2015/3/3-2.go
package main

import (
    "fmt"
)

type loc struct {
    x int
    y int
}

func main() {

    script := getFileLines("input.txt")[0]

    visits := []loc{}

    addVisit(&visits, 0, 0)

    santa := loc{x:0, y:0}
    bot := loc{x:0, y:0}

    for i, c := range script {
        if i % 2 == 0 {
            moveAndVisit(&santa, c, &visits)
        } else {
            moveAndVisit(&bot, c, &visits)
        }
    }
    fmt.Printf("Visits: %v\n", len(visits))
    fmt.Printf("Unique Visits: %v\n", len(uniqueHouses(visits)))

}

func moveAndVisit(current *loc, direction rune, visits *[]loc) {
    switch direction {
    case '<':
        current.x -= 1
    case '>':
        current.x += 1
    case '^':
        current.y += 1
    case 'v':
        current.y -= 1
    default:
        fmt.Printf("Not sure about %v in script\n", direction)
    }
    addVisit(visits, current.x, current.y)
}

func addVisit(visits *[]loc, x int, y int) {
    *visits = append(*visits, loc{x: x, y: y})
}

func uniqueHouses(visits []loc) ([]loc) {
    uniq := []loc{}
    for _, c := range visits {
        already := false
        for _, e := range uniq {
            if c.x == e.x && c.y == e.y {
                already = true
            }
        }
        if already == false {
            uniq = append(uniq, c)
        }
    }
    return uniq
}
